### DISCUSSION TITLE
new ORM statement paradigm does not allow access to methods of class model

### DISCUSSION THREAD

**Discussion by petterroea** [sqlalchemy/sqlalchemy#12015] [Usage Questions] [answered]
Hello all. Working on introducing SQLa into a new project, and wanted to use the new syntax. However, patterns i am used to no longer work. In particular, I am used to using methods in the database model class to handle business logic relating to said class. Turns out this does not work with the new syntax, and this has been a problem since at least 1.4.54.

When using the old syntax(`session.query(A).filter(A.name == "foo").first()`), you can access the methods defined on the class as usual (`a.foo()`). I have attached an example below showing this behavior. However, it seems the object is not aware of its own methods when sqlalchemy has assembled it as a result of running `session.execute`. Instead, if you try to call the method `foo` on class A, you get a trace like this:

```
Traceback (most recent call last):
  File "[redacted]/sqla-regression-mre/venv/lib/python3.12/site-packages/sqlalchemy/engine/result.py", line 199, in _key_not_found
    self._key_fallback(key, None)
  File "[redacted]/sqla-regression-mre/venv/lib/python3.12/site-packages/sqlalchemy/engine/result.py", line 138, in _key_fallback
    raise KeyError(key) from err
KeyError: 'foo'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "[redacted]/sqla-regression-mre/main.py", line 37, in <module>
    run()
  File "[redacted]/sqla-regression-mre/main.py", line 21, in run
    print(selected_a.foo())
          ^^^^^^^^^^^^^^
  File "lib/sqlalchemy/cyextension/resultproxy.pyx", line 66, in sqlalchemy.cyextension.resultproxy.BaseRow.__getattr__
  File "lib/sqlalchemy/cyextension/resultproxy.pyx", line 63, in sqlalchemy.cyextension.resultproxy.BaseRow._get_by_key_impl
  File "[redacted]/sqla-regression-mre/venv/lib/python3.12/site-packages/sqlalchemy/engine/result.py", line 201, in _key_not_found
    raise AttributeError(ke.args[0]) from ke
AttributeError: foo
```

I want this behavior because it lets me treat the objects returned by SQLA as normal python objects that i can design and use using normal object oriented patterns.

Is this expected behavior or a bug? 

# Minimum Reproducable Example

## main.py

```py
from sqlalchemy import create_engine, select
from sqlalchemy.orm import Session
from db import A, Base

def run():
    engine = create_engine("sqlite:///foo.sqlite", echo=True)

    Base.metadata.create_all(engine)

    with Session(engine) as session:
        # Create an A
        a = A("foo")
        session.add(a)

        # Will not work
        #selected_a = session.execute(select(A).where(A.name=="foo")).first()

        # Will work
        selected_a = session.query(A).filter(A.name == "foo").first()
        print(type(selected_a))
        print(selected_a.foo())

        session.commit()


if __name__ == "__main__":
    run()
```

## db.py

```py
from typing import List
from typing import Optional
from sqlalchemy import ForeignKey
from sqlalchemy import String
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import Column
#from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import relationship

from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class A(Base):
    __tablename__ = "a"

    uuid = Column(Integer, primary_key=True)
    name = Column(String)

    def __init__(self, name):
        self.name = name
    
    def foo(self):
        print(f"I am {self.name}")
```

## requirements.txt
Issue is reproducable in at least both of these versions
```
SQLAlchemy==2.0.36
#SQLAlchemy==1.4.54
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

I would suggest you read the migration notes, like https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#orm-session-execute-uses-future-style-result-sets-in-all-cases

Execute will return a result that returns tuple in all cases. You can call scalars() on the results to get the first item on that row. session also has a scalars method, that does that before returning the result

**Reply by petterroea** [reply]
Thank you for your quick response. I understand now.

I had already read the docs, but it was unclear for me how the tuple changes would affect methods defined on the class. In hindsight, the sentence `[...]what this means is that when the [Row](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row) object is returned, the row is a simple tuple that contains the data in its final form,[...]` fully explains the change. 

However, mixed in a document that mainly seems to care about having rows behave more like tuples and less like mappings, it's easy to assume previously existing behavior is untouched. Other sentences like `The base [Row](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row) object now behaves more fully like a named tuple` help enforce this interpretation. Combined with the documentation communicating an intent for the new and old syntax to share the same backend, it was easy for me to assume these changes weren't affecting me. 

I think the documentation can be more clear on practical implications of the new ways things are done. In particular, this section seems to be a good area for improvement, as `scalars` is used without the reasoning being explained: https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#orm-query-is-internally-unified-with-select-update-delete-2-0-style-execution-available

I am more than willing to draft up a PR with some suggestions if that is wanted.

**Reply by CaselIT** [reply]
> The base [Row](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row) object now behaves more fully like a named tuple help enforce this interpretation

The orm classes are not named tuples, so I don't think the docs are suggesting that interpretation.

> I think the documentation can be more clear on practical implications of the new ways things are done. In particular, this section seems to be a good area for improvement, as scalars is used without the reasoning being explained: https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#orm-query-is-internally-unified-with-select-update-delete-2-0-style-execution-available

that section is likely fine as is. that's not the main section that talks about it. That would be https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-orm-usage
